Лесникова Мария

Minniezator opened the pull request.

Minniezator committed:
Update phone-book.js

Minniezator committed:
Update index.js

maxvipon commented on index.js:
```diff
@@ -3,8 +3,8 @@
-phoneBook.add('Сергей', '7 999 6667778', 'gs@example.com');
-phoneBook.add('Сергей 2', '999 4433444', 'gs@example.com');
+phoneBook.add('Сергей', '7 999 666 7 778', 'gs@example.com');
+phoneBook.add('Сергей 2', '999 443 3 444', 'gs@example.com');
```
почему заменила формат этих двух номеров?

Minniezator replied:
Их нет в списке валидности в тз. Изменила на валидные, чтоб в книгу добавились.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
-
-    // Ваша невероятная магия здесь
-
+    if (/^((\+)?(\d{1,4}\s)?(\d{3}|(\(\d{3}\)))\s\d{3}((-\d{1}-)|(\s\d{1}\s))\d{3})|(\d{11,14})$/.test(phone) &&
```
Регулярные выражения лучше вынести в переменные, иначе условие нечитаемо.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
-    // Ваша удивительная магия здесь
-
+    for (var i = 0; i < phoneBook.length; ++i) {
+        for (var key in phoneBook[i]) {
```
У этого способа итерирования по ключам объекта есть недостаток. Поищи и почитай про `hasOwnProperty`

Minniezator replied:
Не поняла, в чем недостаток в данной задаче. Мы же не наследуем ничего.

Minniezator replied:
Вообще до этого вместо <code>for (var key in phoneBook[i])</code> был какой-то такой if
<code>if(phoneBook[name].indexOf(query) > -1 && phoneBook[phone].indexOf(query) > -1 && phoneBook[email].indexOf(query) > -1)</code> :)

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
-
+    for (var i = 0; i < phoneBook.length; ++i) {
+        for (var key in phoneBook[i]) {
+            if (phoneBook[i][key].indexOf(query) > -1) {
```
`.indexOf` чувствителен к регистру, что не очень подходит для поиска. Рекомендую для данной задачи использовать регулярное выражение.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
+    for (var i = 0; i < phoneBook.length; ++i) {
+        for (var key in phoneBook[i]) {
+            if (phoneBook[i][key].indexOf(query) > -1) {
+                for (var key in phoneBook[i]) {
```
_Tip_. Почитай про методы конструктора [`Object`](http://www.google.com/search?q=mdn+object&oq=mdn+obj&sourceid=chrome&ie=UTF-8), там есть то, что позволит тебе упростить код.

maxvipon replied:
После того, как найдешь нужный метод, почитай про итераторы [`Array`](https://www.google.com/search?q=mdn+object&oq=mdn+obj&sourceid=chrome&ie=UTF-8&gws_rd=ssl#newwindow=1&q=mdn+array)

Minniezator replied:
Object.keys(), Array.prototype.forEach()?..

maxvipon replied:
yeap

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
+    var numOfRemoved = 0;
+    for (var i = 0; i < phoneBook.length; ++i) {
+        for (var key in phoneBook[i]) {
+            if (phoneBook[i][key].indexOf(query) > -1) {
```
Все замечания к методу `find` применимы и к этому методу тоже.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
-};
+function createArrayOfMaxLengths() {
+    for (var key in phoneBook[0]) {
+        arrayOfMaxLengths[key] = phoneBook[0][key].length;
```
Название и тип переменной некорректно. Ты используешь массив как объект.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
+    }
+    for (var i = 1; i < phoneBook.length; ++i) {
+        for (var key in phoneBook[i]) {
+            arrayOfMaxLengths[key] = phoneBook[i][key].length > arrayOfMaxLengths[key] ?
```
Тернарный оператор лучше заменить на простой `if`, чтобы не присваивать значение самому себе.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
-*/
-module.exports.showTable = function showTable(filename) {
+function drawHorizontal(lengthOfString) {
+    for (var i = 0; i < lengthOfString + 4 * Object.keys(phoneBook[0]).length; ++i) {
```
Что за магическая чиселка 4?

Minniezator replied:
2 ' ' + 2 '│'
Для каждого столбца добавляется, длина увеличивается на 4.

maxvipon replied:
Такие константы принято именовать, чтобы другим разработчикам было понятней:

``` js
const CELL_PADDING = 4;

// ...

for (var i = 0; i < lengthOfString + CELL_PADDING; i++) { ... }
```

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,101 @@
+        }
+        drawHorizontal(lengthOfString);
+    }
+    return null;
```
возвращать `null` тут некорректно

maxvipon commented:
Поработай над исправлением (а лучше заменой) конструкции `for (var key in object)`.

maxvipon commented:
_JFYI_. Твоя реализация телефонной книги позволяет добавить абсолютно идентичные контакты, что некошерно для телефонной книги))

Minniezator commented:
@maxvipon По поводу идентичных контактов.
Есть способ использовать метод find внутри функций модуля?
Чтоб в add, например, написать:

<pre><code>if (phonePattern.test(phone) && emailPattern.test(email) &&
    find(name) && find(phone) && find(email)) {
        phoneBook.push({name: name, phone: phone, email: email});
    }</code></pre>

Minniezator committed:
Fix_1 phone-book.js

Посмотри find. Я, наверное, что-то не так поняла, потому что код ни разу не упростился. -__-

Minniezator committed:
Fix_2 phone-book.js

В remove тоже убрала for(var key in object). В рисовании таблицы - не стала.

maxvipon commented:
> Есть способ использовать метод find внутри функций модуля?

Сие выражение мне непонятно. Попробуй переформулировать.

> `if (find(name) && find(phone) && find(email)) { ... }`

Всё в твоей власти)

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,106 @@
-
+    var phonePattern = /^((\+)?(\d{1,4}\s)?(\d{3}|(\(\d{3}\)))\s\d{3}((-\d{1}-)|(\s\d{1}\s))\d{3})|(\d{11,14})$/;
+    var emailPattern = /^[a-zA-Z0-9_.+-]+@[a-zA-Zа-яА-Я0-9-]+(\.[a-zA-Zа-яА-Я0-9-.]+)+$/;
+    if (phonePattern.test(phone) && emailPattern.test(email) /*&& !module.exports.find(name) && !module.exports.find(email) && !module.exports.find(email)*/) {
```
Очпросто:

``` js
function checkExistence(name, phone, email) { ... }
function findContact(query) { ... }

function addContact(filename) {
  var exists = checkExistance(fields[0], fields[1], fields[2]);
  // ...
}

exports.add  = addContact;
exports.find = findContact;
```

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,106 @@
+    if (phonePattern.test(phone) && emailPattern.test(email) /*&& !module.exports.find(name) && !module.exports.find(email) && !module.exports.find(email)*/) {
+        phoneBook.push({name: name, phone: phone, email: email});
+    }
+    //else console.log('Данные введены неверно!');
```
А вот это правильно: нужно обязательно сообщить, что была допущена ошибка.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,106 @@
 module.exports.find = function find(query) {
+    var isFound = false;
+    function tryContact(contact, index, book) {
+        function printContact(key, index, keys) {
```
Тут стоит добавить проверку `if (isFound) return` чтобы не продолжать поиск, если искомый контакт уже найден.

Minniezator replied:
Мы же вроде <i>все</i> контакты ищем, удовлетворяющие <code>query</code>.

maxvipon commented:
Вместо ожидаемого

``` bash
❯ node index.js
Сергей, +7 (999) 666-7-778, gs@yandex-team.ru
Олег, +7 (999) 777-7-777, just7@yandex-team.ru
Удален 1 контакт
┌─────────────┬────────────────────╥──────────────────┐
│ Имя         │ Телефон            ║ email            │
├─────────────┼────────────────────╫──────────────────┤
│ Сергей      │ +7 (999) 666-77-78 ║ gs@example.com   │
│ Сергей 2    │ +7 (999) 443-34-44 ║ gs@example.com   │
└─────────────┴────────────────────╨──────────────────┘
```

при запуске твоей программы вижу

``` bash
❯ node index.js
Олег, +7 (999) 777-7-777, just7@yandex-team.ru, ─────────────────────────────────────────────────
| NAME     || PHONE           || EMAIL          |
─────────────────────────────────────────────────
│ Сергей   ││ 7 999 666 7 778 ││ gs@example.com │
│ Сергей 2 ││ 999 443 3 444   ││ gs@example.com │
─────────────────────────────────────────────────
```

То есть метод `find` работает некорректно, метод `remove` не сообщает о результате своей работы.

maxvipon commented:
Советую почитать про то, как работает телефония и что такое по сути телефонный номер, чтобы понять, зачем в нем скобки, дефисы и пробелы, а также знак `+`.

Minniezator committed:
Update phone-book.js

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,127 @@
+            isExist = true;
+        }
+    }
+    phoneBook.forEach(tryContact);
```
_На будущее:_ внимательнее читай, а лучше изучай [документацию](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array)!
Вместо `.forEach()` тут нужно использовать `.some()`.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,127 @@
+        }
+        Object.keys(contact).forEach(isInBook);
+    }
+    phoneBook.forEach(tryContact);
```
_Tip_: тут можно использовать `.find()`, если ты на nodejs v4

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,127 @@
+                console.log();
+            }
+        }
+        Object.keys(contact).forEach(isInBook);
```
_Tip_: вместо проверки каждого ключа, можно сначала сконкатинировать значения всех ключей, а потом проверить регуляркой.

maxvipon commented on phone-book.js:
```diff
@@ -1,52 +1,127 @@
 
+module.exports.showTable = function showTable() {
+    if (phoneBook.length !== 0) {
+        var lengthOfString = createMaxLengths();
```
_Tip_: вычисление самого длинного значания можно производить в момент добавления/удаления контаков

maxvipon commented:
:ok:

maxvipon commented:
:star:
